components/indexer.py

```diff
@@ -1,56 +1,133 @@
 import rev
 import wpilib
+from enum import Enum, auto
 from magicbot import tunable, feedback
-from typing import Tuple
 
 
 class Indexer:
-    indexer_front_motor: rev.CANSparkMax
-    indexer_back_motor: rev.CANSparkMax
+    class Direction(Enum):
+        OFF = 0
+        FORWARDS = 1
+        BACKWARDS = -1
+
+    class CargoColour(Enum):
+        NONE = auto()
+        RED = auto()
+        BLUE = auto()
+
+    # The "tunnel" is the horizontal part of the indexer that the cargo enters first
+    # The "chimney" is the vertical section of the indexer that feeds the shooter
+    # The "cat flap" is the moving flap that allows an opposition cargo to be captured
+    indexer_tunnel_motor: rev.CANSparkMax
+    indexer_chimney_motor: rev.CANSparkMax
     colour_sensor: rev.ColorSensorV3
-    prox_sensor1: wpilib.DigitalInput
-    prox_sensor2: wpilib.DigitalInput
-    # indexer_piston: wpilib.Solenoid
+    lower_chimney_prox_sensor: wpilib.DigitalInput
+    upper_chimney_prox_sensor: wpilib.DigitalInput
+    cat_flap_piston: wpilib.DoubleSolenoid
 
     is_firing = tunable(False)
-    is_red = tunable(False)
-    indexer_speed = tunable(0.5)
-    # Front, back
-    speeds = (0.0, 0.0)
+    tunnel_speed = tunable(0.3)
+    chimney_speed = tunable(1.0)
+
+    _tunnel_direction = Direction.OFF
+    _chimney_direction = Direction.OFF
+    _cat_flap_is_open = False
+
+    last_colour = CargoColour.NONE
+
+    has_trapped_cargo = tunable(False)
 
     def setup(self) -> None:
-        self.indexer_front_motor.setInverted(True)
-        self.indexer_back_motor.setInverted(False)
+        self.indexer_tunnel_motor.setInverted(True)
+        self.indexer_tunnel_motor.setIdleMode(rev.CANSparkMax.IdleMode.kBrake)
+        self.indexer_chimney_motor.setInverted(False)
+        self.indexer_chimney_motor.setIdleMode(rev.CANSparkMax.IdleMode.kBrake)
 
     def execute(self) -> None:
-        self.indexer_front_motor.set(self.speeds[0])
-        self.indexer_back_motor.set(self.speeds[1])
+        if self._tunnel_direction is Indexer.Direction.BACKWARDS:
+            self.indexer_tunnel_motor.set(-1.0)
+        elif self._tunnel_direction is Indexer.Direction.FORWARDS:
+            self.indexer_tunnel_motor.set(self.tunnel_speed)
+        else:
+            self.indexer_tunnel_motor.set(0.0)
+        self.indexer_chimney_motor.set(
+            self.chimney_speed * self._chimney_direction.value
+        )
+        if self._cat_flap_is_open:
+            self.cat_flap_piston.set(wpilib.DoubleSolenoid.Value.kForward)
+        else:
+            self.cat_flap_piston.set(wpilib.DoubleSolenoid.Value.kReverse)
+
+        if self.tunnel_has_red():
+            self.last_colour = Indexer.CargoColour.RED
+        elif self.tunnel_has_blue():
```

Comment:
would be prudent to check whether the sensor thinks it's both red and blue, and reject it